spring-vaadin-security/src/main/java/org/vaadin/spring/security/provider/AbstractAnnotationAccessDelegate.java

```diff
@@ -0,0 +1,166 @@
+
+	private final Logger logger = LoggerFactory.getLogger(getClass());
+	
+	protected ApplicationContext applicationContext;
+	protected VaadinSecurity security;
+	
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		this.applicationContext = applicationContext;		
+	}
+	
+	@Override
+	public void setVaadinSecurity(VaadinSecurity vaadinSecurity) {
+		this.security = vaadinSecurity;
+	}
+	
+	@Override
+	public void afterPropertiesSet() throws Exception {
+		logger.trace("> afterPropertiesSet()");
+		
+		Assert.notNull(applicationContext);
+		Assert.notNull(security);
+		
+		logger.trace("< afterPropertiesSet()");
+	}
+	
+	/**
+	 * Helper class which decides if a object is annotated and configures the access decision manager
+	 * with the correct values.
+	 * 
+	 * @param beanName {@link String} bean name
+	 * @param ui {@link UI} UI instance
+	 * @param annotation Annotation class which to check
+	 * @param annotationAttribute Annotation attribute which contains the arguments for the access decision manager
+	 * 
+	 * @return <code>true</code> when access is granted by the access decision manager, <code>false</code> when access is denied by the access
+	 * decision manager.
+	 */
+	protected boolean isAccessGrantedForAnnotation(String beanName, UI ui, Class<? extends Annotation> annotation, String annotationAttribute) {
+		
+		Annotation annotationClass = applicationContext.findAnnotationOnBean(beanName, annotation);
+		
+		if ( annotationClass == null ) {
+        	logger.trace("{} annotation not present on view", annotation.getSimpleName());
+            return true;
+        } else if ( security.hasAccessDecisionManager() ) {
+
+        	logger.trace("DecisionManager present");
+        	
+            final Class<?> targetClass = AopUtils.getTargetClass(applicationContext.getBean(beanName));
+            final Method method = ClassUtils.getMethod(AopUtils.getTargetClass(applicationContext.getBean(beanName)), "enter", com.vaadin.navigator.ViewChangeListener.ViewChangeEvent.class);
+            final MethodInvocation methodInvocation = MethodInvocationUtils.createFromClass(targetClass, method.getName());
+
+            final Authentication authentication = security.getAuthentication();
+            final AccessDecisionManager accessDecisionManager = security.getAccessDecisionManager();
+            final ExpressionBasedAnnotationAttributeFactory attributeFactory = new ExpressionBasedAnnotationAttributeFactory(new DefaultMethodSecurityExpressionHandler());
+
+            Collection<ConfigAttribute> attributes = new ArrayList<ConfigAttribute>();
+            
+            try {
+            	
+            	logger.trace("Fetching value() from annotation");
+            	Object annotationValue = getAnnotationValue(targetClass, annotation, annotationAttribute);
```

Comment:
Good plan, I will try to come up with a smart way to cache it, let me think about it.